jaxrs/src/test/java/io/airlift/jaxrs/TestRequestScopedBinding.java

```diff
@@ -0,0 +1,141 @@
+import java.util.function.Supplier;
+
+import static io.airlift.http.client.HttpClientBinder.httpClientBinder;
+import static io.airlift.http.client.Request.Builder.prepareGet;
+import static io.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static io.airlift.jaxrs.JaxrsBinder.jaxrsBinder;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Objects.requireNonNull;
+import static org.testng.Assert.assertEquals;
+
+public class TestRequestScopedBinding
+{
+    @Test
+    public void testRequestScopedInjection()
+    {
+        Injector injector = startServer();
+        try {
+            HttpClient httpClient = injector.getInstance(Key.get(HttpClient.class, ForTest.class));
+            URI serverUri = injector.getInstance(HttpServerInfo.class).getHttpUri();
+
+            assertEquals(executeRequest(httpClient, serverUri, "Bob"), "Bob@127.0.0.1");
+            assertEquals(executeRequest(httpClient, serverUri, "Alice"), "Alice@127.0.0.1");
+        }
+        finally {
+            injector.getInstance(LifeCycleManager.class).stop();
+        }
+    }
+
+    private String executeRequest(HttpClient httpClient, URI baseUri, String name)
+    {
+        Request request = prepareGet()
+                .setUri(baseUri.resolve("/hello"))
+                .setHeader("X-Name", name)
+                .build();
+        return httpClient.execute(request, createStringResponseHandler()).getBody();
+    }
+
+    @Path("/hello")
+    public static class MyResource
+    {
+        @GET
+        public String returnUserScopedInfo(@Context RemoteAddr remoteAddr, @Context HeaderValue headerValue)
+        {
+            return headerValue.getValue() + "@" + remoteAddr.value();
+        }
+    }
+
+    @Retention(RUNTIME)
+    @BindingAnnotation
+    public @interface ForTest {}
+
+    public record RemoteAddr(String value) {}
+
+    public static class HeaderValue
+    {
+        private final String value;
+
+        public HeaderValue(String value)
+        {
+            this.value = requireNonNull(value, "value is null");
+        }
+
+        public String getValue()
+        {
+            return value;
+        }
+    }
+
+    private Injector startServer()
+    {
+        return new Bootstrap(
+                binder -> {
+                    jaxrsBinder(binder).bind(MyResource.class);
+                    httpClientBinder(binder).bindHttpClient("test", ForTest.class);
+                    jaxrsBinder(binder).bindRequestScoped(RemoteAddr.class, RemoteAddrFactory.class);
+                    jaxrsBinder(binder).bindRequestScoped(HeaderValue.class, HeaderValueFactory.class);
+                },
+                new TestingNodeModule(),
+                new TestingHttpServerModule(),
+                new JaxrsModule(),
+                new JsonModule())
+                .quiet()
+                .initialize();
+    }
+
+    private record RemoteAddrFactory(HttpServletRequest request)
```

Comment:
How does the injection work if the classes and constructors are private? Does Jersey bypass permissions with reflection?

Reply:
As far as I know - yes